### PR TITLE
Feature: add/remove saved searches.

### PR DESCRIPTION
**Pull request by brunodea**
Since I was already dealing with the subject, I decided to give it a try and implement the save/delete new searches.

here are some images:
http://i.imgur.com/OiK7sPc.png
http://i.imgur.com/XeS1gg0.png?1
http://i.imgur.com/nAxEPFv.png?1

What do you think? Also, it is saying here "Can't automatically merge. Don't worry, you can still create the pull request.", any idea why?

EDIT: I already did ```git rebase develop``` and fixed the conflicts, but it still says here that there are merge conflicts! My git skills are failing me, any help? :-(

$ git rebase ankidroid/develop 
La branche courante feature-add-new-searches est à jour. //current branch feature-add-new-searches is up to date.
$ git push -f origin feature-add-new-searches 
Everything up-to-date

### CODE REVIEW

**Comment by timrae** [comment]
Nice, only thing is I don't really like the button... maybe you can extend one of the classes from `MaterialDialog` and use a builder? 

I'm not sure how you have your remotes setup but I'd normally go:

```
git fetch
git checkout develop
git pull upstream develop
git checkout feature-add-new-searches 
git rebase develop
```

**Comment by brunodea** [comment]
> Nice, only thing is I don't really like the button... maybe you can extend one of the classes from MaterialDialog and use a builder?
Do you mean the "delete" button? Which one should I use? Or are you talking about the "save new search" button?
I did use a MaterialDialog builder to create the dialog in ```onCreateDialog()```, so I didn't really understand what you mean, could you explain a bit further?

About the git commands: I perform them and everything is always up to data :-( I'm seriously thinking about just creating another branch from develop and paste my new code and try again, probably will work. Thanks anyway!

**Comment by hssm** [comment]
The deleting part is fine but the adding part needs work I think.

How about this: instead of having two text boxes for name and query, only have one for name. When saving, use the query that is currently in the browser search bar. That way, the user can write a search string, run the search, figure out that it works, then save it. It would be very inconvenient if they had to type the query again in the new window after they are satisfied that it works.

I agree about the button. It looks really large and out of place.

Let's agree on the commit before we worry about the git stuff for now.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 1]
```diff
@@ -265,17 +265,51 @@ public boolean onSelection(MaterialDialog materialDialog, View view, int which,
...
+                } else if (!savedFiltersObj.has(searchName)) {
+                    savedFiltersObj.put(searchName, searchTerms);
+                    should_save = true;
+                }
+                if (should_save) {
+                    getCol().getConf().put("savedFilters", savedFiltersObj);
+                    getCol().flush();
+                }
+            } catch (JSONException e) {
+            }
```
Comment (hssm) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java]: Can you throw a new RuntimeException for these? It's harder to track down problems if they fail silently like this.

**Comment by brunodea** [comment]
Instead of a button for deleting, what about show a menu (in a dialog form)
after a long press? It would have the options "edit" to edit the query and
"remove".

The adding part  could be then just a button that opens a dialog with the
two edittexts, where the query text box would be filled with the current
user's search so he knows what he is saving and it also allows him to edit.
Or do you prefer just one text box for the name in the same dialog as the
listview? If so, the button for saving could be the same chek mark as for
adding a new tag.

(Of course I can throw a RuntimeException in the catch blocks, it is a good
idea indeed)

**Comment by hssm** [comment]
I think a long press inside a dialog isn't very obvious (people already struggle to figure out the long press on the deck list).

Another idea is to keep the single text box for the name and save the button, but preview the query and name in the confirmation dialog so it's more clear what is happening. I really don't know what the best way to go about it is. We can certainly improve it later on if we have to.

**Comment by timrae** [comment]
I agree about long-press... I think the icons for delete are fine, and edit
feature is probably not really necessary, since it's relatively easy to
just delete and recreate.
Instead of having a Dialog you could make it an Activity with a proper
action bar, with a tick button for accepting the changes. It's up to you
how you implement it, I just hope you can make it fit in properly with the
material design.

**Comment by brunodea** [comment]
Images of how it looks now:
[save search?](http://i.imgur.com/3I7jSZQ.png)
[no saved searches](http://i.imgur.com/sajAWpL.png) (the editbox won't show if there isn't a search query in the SearchView - [here](http://i.imgur.com/hCERS3J.png))
[remove search](http://i.imgur.com/TvxzJPC.png)
[some saved searches and a query in the searchview](http://i.imgur.com/rZuZJes.png)
[some saved searches and no query](http://i.imgur.com/1v6in2S.png)

What do you think? I believe creating an activity just for this isn't really necessary. A Dialog seems good enough.

**Comment by timrae** [comment]
I think the dialog in the second screenshot (and background of the first screenshot) is unnecessary... just add a separate action to the overflow menu -- "save search" which is visible when a search has been entered. Then you can get rid of the TextView entirely.

Also, how about replacing the delete icon with one of the standard material design "remove" icons? For example [this](https://github.com/google/material-design-icons/blob/master/content/drawable-xhdpi/ic_remove_black_24dp.png) or [this](https://github.com/google/material-design-icons/blob/master/content/drawable-xhdpi/ic_remove_circle_black_24dp.png)?

**Comment by timrae** [comment]
Or if you prefer to have the add functionality inside "my searches", maybe just have a separate button which brings up a dialog with two TextViews and standard material design OK and CANCEL buttons... I just think the TextView looks out of place.

**Comment by brunodea** [comment]
[here](http://imgur.com/a/XPcVh#0) is an album with some new images. Both icons for removing are there, which do you prefer?

Now the overflow menu shows "save search" if there is anything written in the SearchView and the "my searches" menu item only is shown in the menu if there are saved searches.

note: I had to update the materialdialog version in the gradle, because the version that we were using didin't have the ```builder.input()``` method.

**Comment by hssm** [comment]
I'm liking this latest version. I prefer the second delete button style. If we agree on the style, can you remove the image that isn't being used?

It should be good to merge after that.

**Comment by timrae** [comment]
Yeah I prefer the second style, but I think the black is a bit heavy, how about we go with the grey one `ic_remove_circle_grey600_36dp`?

**Comment by timrae** [comment]
Also I'd prefer to have the `saveSearchBuilder` as a DialogFragment

**Comment by timrae** [comment]
Sorry one more idea... how about hiding "my searches" from the menu when there are no searches to show? @hssm

**Comment by timrae** [comment]
Do the saved searches work as expected when a deck has been selected? We automatically add "deck: SELECTED_DECK" to searches coming from the search bar, but maybe we should bypass this for saved searches?

**Comment by hssm** [comment]
>Sorry one more idea... how about hiding "my searches" from the menu when there are no searches to show?
It's already doing this.

>Do the saved searches work as expected when a deck has been selected?
No, they are still limiting to the deck. That would have to be fixed as well.

**Comment by timrae** [comment]
> It's already doing this.

Oops, you're right.

> No, they are still limiting to the deck. That would have to be fixed as well.

Yeah, probably any deck limiter should be added explicitly to the saved searches themselves, and when saved searches are recalled the selected deck unselected (i.e. changed to "whole collection"). Hopefully that doesn't introduce any issues when switching back to other screens that relied on the deck being selected properly.

**Comment by brunodea** [comment]
I added the "deck: SELECTED_DECK" in the saved terms. Is that what you requested?

**Comment by timrae** [comment]
@brunodea not quite; the remaining problem is that if you recall a search like "deck:default foo" while you have any deck other than the default selected then no cards will appear.

**Comment by brunodea** [comment]
Ah, ok. Should I try to fix it in this PR? 

Also, Do you want me to remove the "deck: xxx" addition that I made? 

and I've found a problem: wheb the search terms is too large, they overlap the search name and the delete button, so I'll fix it soon first.

**Comment by timrae** [comment]
> Should I try to fix it in this PR?

You're welcome to use a separate PR if you like, it's completely up to you

> Do you want me to remove the "deck: xxx" addition that I made?

I'm not sure whether or not we should include the "hidden" deck filter... What do you think @hssm?

**Comment by brunodea** [comment]
> You're welcome to use a separate PR if you like, it's completely up to you

I think that a new PR makes more sense, since these are different issues.

> I'm not sure whether or not we should include the "hidden" deck filter... What do you think @hssm?

If I had to vote, I would say that we should not include it, because it is easy for the user to just pick the deck and make the search and it also allows the user to make the same search in different decks if he wishes to.

**Comment by timrae** [comment]
OK yeah I agree, remove the deck filter from the search terms and submit a separate PR for the other issue

**Comment by brunodea** [comment]
Ok, now I've fixed the UI problems. Everything seems fine.
Later I'll try to fix the other issue in a another PR.

**Comment by timrae** [comment]
Sorry to keep picking on things so late, but I think a title is [not necessary](http://developer.android.com/guide/topics/ui/dialogs.html#AlertDialog) for the delete confirmation dialog. We could then have one less string to translate, by changing `card_browser_list_my_searches_remove_content` to `Delete \"%1$s\"?`

**Comment by brunodea** [comment]
No problem at all, we must try to do the best things possible :-D, I'll fix it.

**Comment by timrae** [comment]
... and looking again at the strings, `card_browser_list_my_searches_new_search_success` is not used right? can you delete it?  :)

**Comment by brunodea** [comment]
You're right! I just removed it.

**Comment by timrae** [comment]
I just noticed that you copied the wrong image files... `ic_remove_circle_grey600_xxdp`
There should only be one value of xx

**Comment by brunodea** [comment]
So it should be the same image for all folders?

**Comment by timrae** [comment]
It should be the same image name, copied from the appropriate folder in the
icon repository

**Comment by brunodea** [comment]
Alright, sorry for this. What dp should we use?

**Comment by timrae** [comment]
Maybe 24dp is fine, that's the same size as the action bar

**Pull request merged**